app/src/main/java/org/oppia/android/app/player/state/StateFragmentPresenter.kt

```diff
@@ -316,7 +319,11 @@ class StateFragmentPresenter @Inject constructor(
       return
     }
 
+    // mark a checkpoint and then update the UI with the new EphemeralState.
     val ephemeralState = result.getOrThrow()
+    // only mark checkpoint if the current state is either of type PENDING_STATE or TERMINAL_STATE.
+    if (ephemeralState.stateTypeCase != EphemeralState.StateTypeCase.COMPLETED_STATE)
+      markExplorationCheckpoint()
```

Comment:
This function kicks off the saving process in the UI layer. The reason behind this function being in the UI layer is that if we move it to the domain layer, we will be able to start the saving process but we will not be able to return the dataProvider to the UI layer to observe the saving process.

To move this function to the domain layer, I have tried creating a new dataProvider in ExplorationProgressController and using functions combineWith and .transform on it so that I can return the save operation's dataProvider to the UI layer. (The main problem here is we that have to check that the exploration is not in Loading_STATE before we can create a checkpoint` )

But so far nothing seems to work.

One more problem with saving in the UI layer is that we will be re-saving the same checkpoint if the user goes to some previous state and then comes back to the pending state. This is because we are saving every time the exploration goes into VIEWING_STATE and the current state is not a COMPLETED_STATE.

Do you have any suggestions to solve this problem?

Reply:
Created a new PR with an alternate approach to save checkpoints without using ap layer using background dispatchers 

[link](https://github.com/oppia/oppia-android/pull/3403) to the PR with the new approach.

Reply:
Replying to #3408 since I think that approach will be easier to bring to an ideal state.